src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -255,27 +255,83 @@ public void Clear()
         /// <returns>true if item contained; false if not</returns>
         public bool Contains(T item)
         {
-            if (_buckets != null)
+            int[]? buckets = _buckets;
+
+            if (buckets != null)
             {
                 int collisionCount = 0;
-                int hashCode = InternalGetHashCode(item);
                 Slot[] slots = _slots;
-                // see note at "HashSet" level describing why "- 1" appears in for loop
-                for (int i = _buckets[hashCode % _buckets.Length] - 1; i >= 0; i = slots[i].next)
+
+                IEqualityComparer<T>? comparer = _comparer;
+
+                if (comparer == null)
                 {
-                    if (slots[i].hashCode == hashCode && _comparer.Equals(slots[i].value, item))
+                    int hashCode = item == null ? 0 : InternalGetHashCode(item.GetHashCode());
+
+                    if (default(T)! != null) // TODO-NULLABLE: default(T) == null warning (https://github.com/dotnet/roslyn/issues/34757)
                     {
-                        return true;
+                        // see note at "HashSet" level describing why "- 1" appears in for loop
+                        for (int i = buckets[hashCode % buckets.Length] - 1; i >= 0; i = slots[i].next)
+                        {
+                            if (slots[i].hashCode == hashCode && EqualityComparer<T>.Default.Equals(slots[i].value, item))
```

Comment:
It's unfortunate that this code block is repeated 3 times but I guess there is no better way to do it.

Maybe there's a way to somehow merge the case where `_comparer == null && default(t) == null` with `_comparer != null` by sharing the `comparer` local, something like this?

```cs
IEqualityComparer<T>? comparer = _comparer != null ? _comparer : default(T) == null ? EqualityComparer<T>.Default : null;

if (comparer == null)
{
    ... // devirtualization code path.
}
else
{
   ... // comparer code path.
}
```

However you can validate with your benchmarks if this affects performance.

Reply:
The code uses the same pattern as in `Dictionary<,>` and I intentionally keep the existing implementation detail in `HashSet` in this change since it's just step one. There're more back-port and further optimization can be done later.

In term of your suggestion here, it's actually not quite the same, since in original change, the 2nd the 3rd branch are not the same. The 2nd branch is using `EqualityComparer<T>.Default` which is a type of `EqualityComparer<T>` (an abstract class), but the 3rd branch is using `IEqualityComparer<T>` (an interface). I've tried this way before and it turns out the performance it a bit slower, probably because virtual method dispatching is faster than interface  method dispatching.

I've also tried to check if the custom `_comparer` is an `EqualityComparer<T>` sub-class (so it becomes four branches), but there's no performance gain. I'll run benchmark again to confirm the result.